gadget/update.go

```diff
@@ -904,6 +905,247 @@ func buildNewVolumeToDeviceMapping(old GadgetData, laidOutVols map[string]*LaidO
+				// just need to find what partition is associated with this
+				// structure to find it's root mount points. On GPT since
+				// partition labels/names are unique in the partition table, we
+				// could do a lookup by matching partition label, but this won't
+				// work on MBR which doesn't have such a concept, so instead we
+				// use the start offset to locate which disk partition this
+				// structure is equal to.
+
+				partitions, err := disk.Partitions()
+				if err != nil {
+					return nil, err
+				}
+
+				var foundP disks.Partition
+				found := false
+				for _, p := range partitions {
+					if p.StartInBytes == uint64(structStartOffset) {
+						foundP = p
+						found = true
+						break
+					}
+				}
+				if !found {
+					dieErr := fmt.Errorf("cannot locate structure %d on volume %s: no matching start offset", volYamlIndex, volName)
+					if fatal := maybeDieOnMappingError(dieErr); fatal {
+						return nil, dieErr
+					} else {
+						return nil, errSkipUpdateProceedRefresh
+					}
+				}
+
+				// if this structure is an encrypted one, then we can't just
+				// get the root mount points for the device node, we would need
+				// to find the decrypted mapper device for the encrypted device
+				// node and then find the root mount point of the mapper device
+				if _, ok := diskDeviceTraits.StructureEncryption[volStruct.Name]; ok {
+					logger.Noticef("gadget asset update for assets on encrypted partition %s unsupported", volStruct.Name)
+
+					// setting this structure as having an empty location will
+					// mean when an update to this structure is actually
+					// performed it will fail, but we won't fail updates to
+					// other structures - it is treated like an unmounted
+					// partition
+					volumeStructureToLocation[volName][volYamlIndex] = StructureLocation{}
+					continue
+				}
+
+				// otherwise normal unencrypted filesystem, find the rw mount
+				// points
+				mountpts, err := disks.MountPointsForPartitionRoot(foundP, map[string]string{"rw": ""})
+				if err != nil {
+					dieErr := fmt.Errorf("cannot locate structure %d on volume %s: error searching for root mount points: %v", volYamlIndex, volName, err)
+					if fatal := maybeDieOnMappingError(dieErr); fatal {
+						return nil, dieErr
+					} else {
+						return nil, errSkipUpdateProceedRefresh
+					}
+				}
+				var mountpt string
+				if len(mountpts) == 0 {
+					// this filesystem is not already mounted, we probably
+					// should mount it in order to proceed with the update?
+
+					// TODO: do something better here?
+					logger.Noticef("structure %d on volume %s (%s) is not mounted read/write anywhere to be able to update it", volYamlIndex, volName, foundP.KernelDeviceNode)
+				} else {
+					// use the first one, it doesn't really matter to us
+					// which one is used to update the contents
+					mountpt = mountpts[0]
+				}
+				volumeStructureToLocation[volName][volYamlIndex] = StructureLocation{
```

Comment:
nitpick: we could have a var StructureLocation before the if and set values on it and move the map assignment after the if

Reply:
done